### ISSUE TITLE
WinCon issues in Windows 11 + MSYS2 Mingw64

### ISSUE THREAD

**Issue by HunterZ** [Bill-Gray/PDCursesMod#235]
Splitting WinCon discussion from issue #234, per following suggestion:

> When you say it was hard to get WinCon working : it shouldn't be any more difficult than WinGUI. Perhaps an issue for that as well? Generally speaking, I'd suggest WinGUI over WinCon; the latter is quite limited in comparison. But for one reason or another, a lot of people still seem to like WinCon, and we do try to keep it working.

So I ran into a number of issues with WinCon, which I will detail below. I don't expect that all or maybe even any of them are your fault, versus MSYS2 package maintainers, Windows 11 Terminal app quirks, etc., but I thought you might appreciate the notes at least:

First, the MSYS2 Mingw64 static library for some reason wouldn't link due to a `PlaySound` symbol error, so I had to switch to the DLL version. Confusingly, statically linking against `libpdcurses_wingui.a` has the same issue, even though it's the exact same file as `libpdcurses.a`, which itself works fine when linked in the exact same manner. This is probably all MSYS2's fault somehow.

After I waded through all that, I found that it seems to be impossible to _programmatically_ resize the new Windows 11 Terminal app, so my output was all scrambled due to being based on very old code that assumes 80 columns. I even tried copy-pasting some Win32 API code, and it had no effect. Is there some way to get WinCon to create its own Console Host window instead of trying to use whatever window it was launched from?

When I did manage to get a correct-looking terminal size, some of my unicode characters still weren't showing up. I tried the non-Mod version of PDCurses and had the same problem there too, though, so it's probably somehow my fault or a problem with the new Windows 11 Terminal app. These characters do show up fine in WinGui mode, however. Maybe I just need to try some different fonts?

After all this, I decided that WinCon is just too fragile for my needs, at least on Windows 11. It's a shame, because unlike WinGui it does support beeps and shift+keypad.

**Comment by GitMensch** [comment]
> Is there some way to get WinCon to create its own Console Host window instead of trying to use whatever window it was launched from?

No, only if you start a sub-process with explicit `cmd /c start cmd /k`.

> After I waded through all that, I found that it seems to be impossible to _programmatically_ resize the new Windows 11 Terminal app, so my output was all scrambled due to being based on very old code that assumes 80 columns.

I can second that, also happens on Windows 10 - works with "free standing" cmd.exe/powershell.exe, doesn't work if cmd.exe/powershell.exe runs below WindowsTerminal.exe.
Also the initial size as reported to the application is wrong, so "bigger" consoles look scrambled (works with parts that are redrawn later) - this and the resize issue can be seen with testcurs.

Using the vt variant doesn't help, both issues are similar there.

Updating to newest [Windows Terminal Preview](https://github.com/microsoft/terminal/releases9 (currently v1.13.10395.0) made the "additional vt issues" a bit better, but did not changed much in general for both ports.

> the MSYS2 Mingw64 static library for some reason wouldn't link due to a `PlaySound` symbol error

For WinCon you need to explicit link against `-lwinmm`.
I think this should be true for both the static and the shared version.
 Did we miss to mention that in the port docs? Does it work when you do that?

**Comment by HunterZ** [comment]
> For WinCon you need to explicit link against `-lwinmm`. I think this should be true for both the static and the shared version. Did we miss to mention that in the port docs? Does it work when you do that?

Could have sworn I tried that, but you're right - it absolutely works. Thanks!

Also verified that my missing characters issue was due to Windows Terminal defaulting to an apparently lackluster font in terms of unicode support.

Discovered a keyboard quirk with WinCon: `PDC_get_key_modifiers()` returns zero when typing on the number pad with numlock enabled. This blocks any ability to differentiate whether the user entered a number via the keypad or the regular number row.

**Comment by GitMensch** [comment]
@HunterZ seems only the last part was still open - can you please recheck if this is still the case, and if so move this out to new issue?

**Comment by Bill-Gray** [comment]
Just tried it with the 'input test' part of `testcurs`.  To do so,  you have to edit line 18 to enable the 'classic mouse interface'.  You shouldn't have to do that;  I'll set about fixing it.  It was an unintended side effect of extending `testcurs` so that both the classic and ncurses mouse interfaces could be tested.

Anyway.  Once you switch,  key modifiers are shown as you type,  including NumLock,  on platforms supporting such a thing.  That includes WinCon,  WinGUI,  and x11new,  but not VT or DOS. [*] 

On Wine,  NumLock appears to go undetected in WinCon.  But that appears to be a Wine quirk;  it's fine on my "real" Win10 box.

Bottom line,  though,  is that it appears this issue is now an ex-issue and can be closed.

[*] On those platforms,  some modifiers are inferred,  but not always accurately.  If you turn on CapsLock and hit A,  it's assumed the shift key is held.  Shift+a and Caps Lock+a return the same key codes to VT and DOS.

Similarly,  DOS and VT can't tell the difference between Shift+(numeric keypad)4,  NumLock+(numeric keypad)4,  and a "normal" 4.  All three send a '4' (ASCII 52) to PDCursesMod.
